### PR TITLE
Add mDNS provider option using hashicorp/mdns library

### PR DESCRIPTION
**Pull request by leprechau**
Adding an mDNS provider option for us on a private lan segment.  I plan to leverage this in an upcoming PR to add mDNS advertisements to consul agents in server mode via the same library.

### CODE REVIEW

**Comment by leprechau** [comment]
Any thoughts on this?

**Comment by pearkes** [comment]
This is interesting, @leprechau! I think the stated goal of go-discover is to support specifically _cloud_ environments, where I believe mdns is not usually possible. Conceptually though, it does seem similar to me so I get the addition. I will try to track someone down to give you a more authoritative answer on the subject!

**Comment by leprechau** [comment]
@pearkes I get the cloud focus, however this is also the discover library used within Consul.  I think having the ability for Consul nodes to discover each other via mDNS on a lan segment or within an isolated Docker network could be very useful.

**Comment by pearkes** [comment]
Okay @leprechau -- after a bit of discussion within HashiCorp I think we should figure out how to get this in 😄. Do you think integration testing similar to how we have for the cloud providers would make sense here?

https://github.com/hashicorp/go-discover/tree/master/test/tf

Another housekeeping note, it'd be great to update the readme with the mDNS provider listed as part of this. We'll come back around for a formal review.

**Comment by leprechau** [comment]
@pearkes I updated the readme to document the mDNS option.

To the integration tests, I thought about that when I first built the PR.  It should be possible to spin up a docker container running an mDNS server and exec `discover` within that container to pull back a service record.  I wasn't sure if that would be sufficient or beneficial though.

**Comment by leprechau** [comment]
Ideally I'd like to see this merged and then submit another PR to Consul to allow it to publish it's own mDNS records.  We could then build an integration test that spun up consul in a container and have `discover` pull the service record advertised by Consul.

**Comment by pearkes** [comment]
@leprechau fair, it'd be great to follow the pattern we have currently of being able to test it independent of HashiCorp tools (I can imagine [Nomad](https://www.nomadproject.io/) wanting to integrate this as well as Consul!). If we can figure out a decent way that'd be great.

In the meantime we'll get a review done here!

**Comment by leprechau** [comment]
@pearkes thank you.  I added a simple integration test into the provider package.  It now validates config and spins up a test local mDNS server and then uses the provider to query for the service.

File: provider/mdns/mdns_provider_test.go [hunk 1]
```diff
@@ -85,9 +117,16 @@ func TestAddrs(t *testing.T) {
 		},
 	}
 
-	p := &mdns.Provider{}
+	p := &provider.Provider{}
 	l := log.New(os.Stderr, "", log.LstdFlags)
 
+	svr, err := newTestServer()
+	if err != nil {
+		t.Fatalf("Failed to start test server: %s", err.Error())
```
[the remark this answers is not in the document]
Comment (leprechau) [reply provider/mdns/mdns_provider_test.go]: `err.Error()` returns a string ... you could use `%v` with just `err` but I don't really see what the difference is in a test case.

**Comment by leprechau** [comment]
@pearkes any updates on getting this merged? If the above comment is the only feedback on the overall PR I consider it done.

**Comment by leprechau** [comment]
@pearkes Thank you sir, appreciate the update.

**Comment by smurfix** [comment]
Any progress? there have been a lot of "couple of" weeks since May 1st … ;-)

**Comment by leprechau** [comment]
I'm still interested in getting this in and adding mDNS server/broadcast support to Consul in a follow-up PR once we have the go-discover side done.

**Comment by euskadi31** [comment]
Any progress?

**Comment by leprechau** [comment]
@pearkes Anything I can do to help get this moved along?

**Comment by leprechau** [comment]
@alvin-huang are you supporting go-discover now?

**Pull request merged**